sns/parity_math.c

```diff
@@ -385,85 +768,911 @@ static void xor_calculate(struct m0_parity_math *math,
+		dest_frags[i] = (uint8_t *)dest_bufs[i].b_addr;
+	}
+
+	ec_encode_data_update(block_size, data_nr, dest_nr, vec_idx,
+			      g_tbls, (uint8_t *)src_buf->b_addr, dest_frags);
+
+	M0_LEAVE();
+}
+
+static void fails_sort(struct m0_reed_solomon *rs, uint8_t *fail,
+		       uint32_t total_count, uint32_t parity_count)
+{
+	uint32_t  i;
+	uint32_t  alive_nr;
+
+	M0_ENTRY();
+
+	M0_PRE(fail != NULL);
+	M0_PRE(rs != NULL);
+	M0_PRE(rs->rs_failed_idx != NULL);
+	M0_PRE(rs->rs_alive_idx != NULL);
+
+	rs->rs_failed_nr = 0;
+	for (i = 0, alive_nr = 0; i < total_count; i++) {
+		if (fail[i] != 0) {
+			VALUE_ASSERT_INFO(rs->rs_failed_nr < parity_count,
+					  rs->rs_failed_nr);
+			rs->rs_failed_idx[rs->rs_failed_nr++] = i;
+		}
+		else
+			rs->rs_alive_idx[alive_nr++] = i;
+	}
+	M0_LEAVE();
+}
+
+static void buf_sort(struct m0_reed_solomon *rs, uint32_t total_count,
+		     uint32_t data_count, uint8_t *fail,
+		     struct m0_buf *data, struct m0_buf *parity)
+{
+	uint32_t  i;
+	uint32_t  j;
+	uint32_t  k;
+	uint8_t  *addr;
+
+	M0_ENTRY();
+
+	M0_PRE(rs != NULL);
+	M0_PRE(fail != NULL);
+	M0_PRE(data != NULL);
+	M0_PRE(parity != NULL);
+	M0_PRE(rs->rs_bufs_in != NULL);
+	M0_PRE(rs->rs_bufs_out != NULL);
+
+	for (i = 0, j = 0, k = 0; i < total_count; i++) {
+		if (i < data_count)
+			addr = (uint8_t *)data[i].b_addr;
+		else
+			addr = (uint8_t *)parity[i - data_count].b_addr;
+
+		if (fail[i] != 0)
+			rs->rs_bufs_out[j++] = addr;
+		else if (k < data_count)
+			rs->rs_bufs_in[k++] = addr;
+		else
+			continue;
+	}
+	M0_LEAVE();
+}
+
+static int isal_gen_recov_coeff_tbl(uint32_t data_count, uint32_t parity_count,
+				    struct m0_reed_solomon *rs)
+{
+	uint32_t  total_count;
+	uint32_t  mat_size;
+	uint32_t  i;
+	uint32_t  j;
+	uint32_t  r;
+	uint8_t   s;
+	uint8_t   idx;
+	int       ret;
+
+	M0_ENTRY("data_count=%u, parity_count=%u, rs=%p",
+		 data_count, parity_count, rs);
+
+	M0_PRE(rs != NULL);
+
+	total_count = data_count + parity_count;
+	mat_size = total_count * data_count;
+
+	uint8_t   decode_mat[mat_size];
+	uint8_t   temp_mat[mat_size];
+	uint8_t   invert_mat[mat_size];
+
+	/* Construct temp_mat (matrix that encoded remaining frags)
+	 * by removing erased rows. */
+	for (i = 0; i < data_count; i++) {
+		idx = rs->rs_alive_idx[i];
+		for (j = 0; j < data_count; j++)
+			temp_mat[data_count * i + j] =
+				rs->rs_encode_matrix[data_count * idx + j];
+	}
+
+	/* Invert matrix to get recovery matrix. */
+	ret = gf_invert_matrix(temp_mat, invert_mat, data_count);
+	if (ret != 0)
+		return M0_ERR_INFO(ret, "failed to construct an %u x %u "
+				   "inverse of the input matrix",
+				   data_count, data_count);
+
+	/* Create decode matrix. */
+	for (r = 0; r < rs->rs_failed_nr; r++) {
+		idx = rs->rs_failed_idx[r];
+		/* Get decode matrix with only wanted recovery rows */
+		if (idx < data_count) {    /* A src err */
+			for (i = 0; i < data_count; i++)
+				decode_mat[data_count * r + i] =
+					invert_mat[data_count * idx + i];
+		}
```

Comment:
[Style] Keep else in the same line as that of closing braces.

Reply:
Done.